src/psyclone/tests/psyir/nodes/dynamic_omp_task_directive_test.py

```diff
@@ -0,0 +1,3915 @@
+        tree.lower_to_language_level()
+    assert ("'ArrayReference' object is not allowed to appear in an array "
+            "index expression inside an OMPTaskDirective."
+            in str(excinfo.value))
+
+
+def test_omp_task_directive_3(fortran_reader, fortran_writer):
+    '''Test the code generation correctly captures if a variable should be
+    declared as firstprivate.'''
+    code = '''
+    subroutine my_subroutine()
+        integer, dimension(10, 10) :: A
+        integer, dimension(10, 10) :: B
+        integer :: i
+        integer :: j
+        integer :: k
+        do i = 1, 10
+            k = i
+        end do
+        do i = 1, 10
+            do j = 1, 10
+                A(i, j) = k
+                A(i, j) = B(i, j) + k
+            end do
+        end do
+    end subroutine
+    '''
+    tree = fortran_reader.psyir_from_source(code)
+    ptrans = OMPParallelTrans()
+    strans = OMPSingleTrans()
+    tdir = DynamicOMPTaskDirective()
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[1]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=1)
+    strans.apply(parent.children[1])
+    ptrans.apply(parent.children)
+    correct = '''subroutine my_subroutine()
+  integer, dimension(10,10) :: a
+  integer, dimension(10,10) :: b
+  integer :: i
+  integer :: j
+  integer :: k
+
+  !$omp parallel default(shared), private(i,j,k)
+  do i = 1, 10, 1
+    k = i
+  enddo
+  !$omp single
+  !$omp task private(i,j), firstprivate(k), shared(a,b), depend(in: b(:,:)), \
+depend(out: a(:,:))
+  do i = 1, 10, 1
+    do j = 1, 10, 1
+      a(i,j) = k
+      a(i,j) = b(i,j) + k
+    enddo
+  enddo
+  !$omp end task
+  !$omp end single
+  !$omp end parallel
+
+end subroutine my_subroutine\n'''
+    assert fortran_writer(tree) == correct
+
+
+def test_omp_task_directive_4(fortran_reader, fortran_writer):
+    ''' Test the code generation correctly makes the depend clause when
+    accessing an input array shifted by the step size of the outer loop.'''
+    code = '''
+    subroutine my_subroutine()
+        integer, dimension(10, 10) :: A
+        integer, dimension(11, 10) :: B
+        integer :: i
+        integer :: j
+        integer :: k
+        do i = 1, 10
+            do j = 1, 10
+                A(i, j) = k
+                A(i, j) = B(i+1, j) + k
+            end do
+        end do
+    end subroutine
+    '''
+    tree = fortran_reader.psyir_from_source(code)
+    ptrans = OMPParallelTrans()
+    strans = OMPSingleTrans()
+    tdir = DynamicOMPTaskDirective()
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[0].children[3].children[0]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=0)
+    strans.apply(loops[0])
+    ptrans.apply(loops[0].parent.parent)
+    correct = '''subroutine my_subroutine()
+  integer, dimension(10,10) :: a
```

Comment:
Removed all the code outside the task directive.

Reply:
I will do this throughout the file.

Reply:
Some tests have a bit more than just the task directive region as I am also checking the data sharing clauses on the parallel directive match (just for a bit extra sanity checking). I also will leave the full task directive region so its clear from a glance that the behaviour of the test makes sense.

Reply:
This will also mean changes to parallel directive's data sharing clauses will more obviously show if/why they break these tests.

Reply:
Great. In doing my other big PR I've also realised that it's nice if tests don't break because of some change that is unrelated to their functionality so it's well worth doing this.